Assets/Scripts/Algorithms/Patrolling/Components/CollisionRecoverySenseNeighboringRobotsComponent.cs

```diff
@@ -0,0 +1,125 @@
+// Jakob Meyer Olsen
+// Mads Beyer Mogensen
+using System.Collections.Generic;
+using System.Text;
+
+using Maes.Map.PathFinding;
+using Maes.Robot;
+using Maes.Robot.Tasks;
+
+namespace Maes.Algorithms.Patrolling.Components
+{
+    internal class CollisionMessage
+    {
+        public CollisionMessage(int myId, Queue<PathStep> currentPath)
+        {
+            RobotId = myId;
+            PathSteps = currentPath;
+        }
+
+        public int RobotId { get; }
+        public Queue<PathStep> PathSteps { get; }
+    }
+    public class CollisionRecoverySenseNeighboringRobotsComponent : IComponent
+    {
+        private const float RecoveryAngle = 45f; // All robots rotate by this angle
+        private readonly Robot2DController _controller;
+        private bool _doingCollisionRecovery;
+
+        public int PreUpdateOrder => -100;
+        public int PostUpdateOrder => -100;
+
+        public CollisionRecoverySenseNeighboringRobotsComponent(Robot2DController controller)
+        {
+            _controller = controller;
+        }
+        public IEnumerable<ComponentWaitForCondition> PreUpdateLogic()
+        {
+            while (true)
+            {
+                if (_controller.IsCurrentlyColliding)
+                {
+                    _doingCollisionRecovery = true;
+                    _controller.StopCurrentTask();
+                    yield return ComponentWaitForCondition.WaitForRobotStatus(RobotStatus.Idle, false);
+
+                    GetNearbyRobots(out var myId, out var lowestId);
+
+                    if (lowestId == myId)
+                    {
+                        foreach (var condition in PerformRecoveryManeuver())
+                        {
+                            yield return condition;
+                        }
+
+                        while (lowestId == myId)
+                        {
+                            yield return ComponentWaitForCondition.WaitForLogicTicks(1, false);
+                            GetNearbyRobots(out myId, out lowestId);
+                        }
+                    }
+                }
+
+                yield return ComponentWaitForCondition.WaitForLogicTicks(1, true);
+                _doingCollisionRecovery = false;
+            }
+        }
+
+        public void DebugInfo(StringBuilder stringBuilder)
+        {
+            if (_doingCollisionRecovery)
+            {
+                stringBuilder.Append("Doing multi-robot collision avoidance\n");
+            }
+        }
+
+        private IEnumerable<ComponentWaitForCondition> PerformRecoveryManeuver()
+        {
+            _controller.Rotate(RecoveryAngle);
+            yield return ComponentWaitForCondition.WaitForRobotStatus(RobotStatus.Idle, false);
+            _controller.Move(2.5f, reverse: true);
+            yield return ComponentWaitForCondition.WaitForRobotStatus(RobotStatus.Idle, false);
+        }
+
+        private void GetNearbyRobots(out int myId, out int lowestId)
```

Comment:
Maybe just return the lowest id. It is easy to get our own id with _controller.RobotId

Reply:
This function is also badly named. It does not get nearby robots.

Reply:
You could also return a bool instead. It is only used for myId == lowestId.

Reply:
I have refactored as i like, returning the lowest id, or -1 if no robots is near